File: 155. Min Stack.py
class MinStack:

    def __init__(self):
        self.res = [] 
        self.minres = []       

    def push(self, val: int) -> None:
        self.res.append(val)
        if self.minres:
            val = min(val, self.minres[-1])
        self.minres.append(val)
        

    def pop(self) -> None:
        self.res.pop()
        self.minres.pop()
        

    def top(self) -> int:
        return self.res[-1]
        

    def getMin(self) -> int:
        return self.minres[-1]
        
        
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: 155. Min Stack.go
type MinStack struct {
    vals []int
    min_val int    
}


func Constructor() MinStack {
    return MinStack{
        vals: []int{},
        min_val: 1<<63-1,
    }
    
}


func (this *MinStack) Push(val int)  {
    if this.min_val > val {
        this.min_val = val
    }
    this.vals = append(this.vals, val)
    
}


func (this *MinStack) Pop()  {
    temp := this.vals[len(this.vals)-1]
    this.vals = this.vals[:len(this.vals)-1]
    if this.min_val == temp {
        this.min_val = 1<<63-1
        for _,i := range(this.vals) {
            if i < this.min_val {
                this.min_val = i
            }
        }
    }

}


func (this *MinStack) Top() int {
    return this.vals[len(this.vals)-1]
    
}


func (this *MinStack) GetMin() int {
    return this.min_val    
}


/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(val);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */